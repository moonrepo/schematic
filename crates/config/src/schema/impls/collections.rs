// use crate::config::ConfigSchema;
// use crate::schema::types::*;

// impl<T: ConfigSchema> ConfigSchema for Vec<T> {
//     fn generate_schema() -> Schema {
//         let inner = T::generate_schema();

//         if let Schema::Type {
//             ref mut nullable, ..
//         } = schema
//         {
//             *nullable = true;
//         }

//         schema
//     }
// }
