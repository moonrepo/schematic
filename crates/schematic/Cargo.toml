[package]
name = "schematic"
version = "0.15.1"
edition = "2021"
license = "MIT"
description = "A layered serde configuration and schema library."
homepage = "https://moonrepo.github.io/schematic"
repository = "https://github.com/moonrepo/schematic"
readme = "../../README.md"

[package.metadata.release]
pre-release-replacements = [
	{ file = "../../CHANGELOG.md", search = "Unreleased", replace = "{{version}}" },
]

[package.metadata.docs.rs]
all-features = true

[dependencies]
schematic_macros = { version = "0.15.0", path = "../macros" }
schematic_types = { version = "0.7.0", path = "../types" }
miette = { workspace = true }
thiserror = "1.0.59"
tracing = "0.1.40"

# config
garde = { version = "0.18.0", default-features = false, optional = true, features = [
	"regex",
] }
serde = { workspace = true }
serde_path_to_error = { version = "0.1.16", optional = true }
starbase_styles = { version = "0.4.0", optional = true }

# schema
indexmap = { workspace = true, optional = true, features = ["serde"] }

# json
serde_json = { workspace = true, optional = true, features = [
	"preserve_order",
] }

# json schema
markdown = { version = "1.0.0-alpha.17", optional = true }
schemars = { version = "0.8.17", optional = true, default-features = false }

# toml
toml = { workspace = true, optional = true }

# yaml
serde_yaml = { workspace = true, optional = true }

# url
reqwest = { workspace = true, optional = true, features = ["blocking"] }

[features]
default = ["config", "url"]
config = [
	"dep:garde",
	"dep:serde_path_to_error",
	"dep:starbase_styles",
	"schematic_macros/config",
]
json = ["dep:serde_json"]
schema = ["dep:indexmap", "schematic_macros/schema"]
toml = ["dep:toml"]
url = ["dep:reqwest"]
yaml = ["dep:serde_yaml"]

renderer_json_schema = ["dep:markdown", "dep:schemars", "json", "schema"]
renderer_template = []
renderer_typescript = ["schema"]

type_chrono = ["schematic_types/chrono"]
type_indexmap = ["schematic_types/indexmap"]
type_regex = ["schematic_types/regex"]
type_relative_path = ["schematic_types/relative_path"]
type_rust_decimal = ["schematic_types/rust_decimal"]
type_semver = ["schematic_types/semver"]
type_serde_json = ["schematic_types/serde_json"]
type_serde_toml = ["schematic_types/serde_toml"]
type_serde_yaml = ["schematic_types/serde_yaml"]
type_url = ["schematic_types/url"]

valid_email = ["garde/email"]
valid_url = ["garde/url"]

[dev-dependencies]
schematic = { path = ".", features = [
	"config",
	"json",
	"schema",
	"toml",
	"renderer_json_schema",
	"renderer_template",
	"renderer_typescript",
	"type_chrono",
	"type_indexmap",
	"type_regex",
	"type_relative_path",
	"type_rust_decimal",
	"type_semver",
	"type_serde_json",
	"type_serde_toml",
	"type_serde_yaml",
	"type_url",
	"url",
	"valid_email",
	"valid_url",
	"yaml",
] }
reqwest = { workspace = true, features = [
	"blocking",
	"rustls-tls-native-roots",
] }
serial_test = "3.1.1"
starbase_sandbox = "0.6.0"

# Types
chrono = { workspace = true }
indexmap = { workspace = true }
regex = { workspace = true }
relative-path = { workspace = true, features = ["serde"] }
rust_decimal = { workspace = true }
semver = { workspace = true, features = ["serde"] }
url = { workspace = true, features = ["serde"] }
