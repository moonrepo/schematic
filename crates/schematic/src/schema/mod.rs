mod generator;
mod renderer;
mod renderers;

pub use generator::*;
pub use indexmap::*;
pub use renderer::*;
pub use schematic_types::*;

/// Renders JSON schemas.
#[cfg(feature = "renderer_json_schema")]
pub use renderers::json_schema::{self, *};

/// Renders JSON config templates.
#[cfg(all(feature = "renderer_template", feature = "json"))]
pub use renderers::json_template::*;

/// Renders JSONC config templates.
#[cfg(all(feature = "renderer_template", feature = "json"))]
pub use renderers::jsonc_template::*;

/// Helpers for config templates.
#[cfg(feature = "renderer_template")]
#[allow(deprecated)]
pub use renderers::template::{self, TemplateOptions, TemplateRenderer};

/// Renders TOML config templates.
#[cfg(all(feature = "renderer_template", feature = "toml"))]
pub use renderers::toml_template::*;

/// Renders TypeScript types.
#[cfg(feature = "renderer_typescript")]
pub use renderers::typescript::{self, *};

/// Renders YAML config templates.
#[cfg(all(feature = "renderer_template", feature = "yaml"))]
pub use renderers::yaml_template::*;
