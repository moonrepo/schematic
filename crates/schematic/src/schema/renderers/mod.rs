#[cfg(feature = "renderer_json_schema")]
pub mod json_schema;

#[cfg(all(feature = "renderer_template", feature = "json"))]
pub mod json_template;

#[cfg(all(feature = "renderer_template", feature = "json"))]
pub mod jsonc_template;

#[cfg(feature = "renderer_template")]
pub mod template;

#[cfg(all(feature = "renderer_template", feature = "toml"))]
pub mod toml_template;

#[cfg(feature = "renderer_typescript")]
pub mod typescript;

#[cfg(all(feature = "renderer_template", feature = "yaml"))]
pub mod yaml_template;
