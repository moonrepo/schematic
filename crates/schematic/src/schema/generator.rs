use super::SchemaRenderer;
use indexmap::IndexMap;
use miette::IntoDiagnostic;
use schematic_types::*;
use std::collections::HashSet;
use std::fs;
use std::path::Path;

/// A generator collects [`Schema`]s and renders them to a specific file,
/// using a renderer that implements [`SchemaRenderer`].
#[derive(Debug, Default)]
pub struct SchemaGenerator<'gen> {
    references: HashSet<String>,
    schemas: IndexMap<String, Schema>,

    _marker: std::marker::PhantomData<&'gen ()>,
}

impl<'gen> SchemaGenerator<'gen> {
    /// Add a [`Schema`] to be rendered, derived from the provided [`Schematic`].
    pub fn add<T: Schematic>(&mut self) {
        let schema = SchemaBuilder::build_root::<T>();
        self.add_schema(&schema, false);
    }

    /// Add an explicit [`Schema`] to be rendered, and recursively add any nested schemas.
    /// Schemas with a name will be considered a reference.
    pub fn add_schema(&mut self, schema: &Schema, is_field: bool) {
        let mut schema = schema.to_owned();

        // Recursively add any nested schema types
        match &mut schema.ty {
            SchemaType::Array(inner) => {
                self.add_schema(&inner.items_type, false);
            }
            SchemaType::Object(inner) => {
                self.add_schema(&inner.key_type, false);
                self.add_schema(&inner.value_type, false);
            }
            SchemaType::Struct(inner) => {
                for field in inner.fields.values() {
                    self.add_schema(field, true);
                }
            }
            SchemaType::Tuple(inner) => {
                for item in &inner.items_types {
                    self.add_schema(item, false);
                }
            }
            SchemaType::Union(inner) => {
                for variant in &inner.variants_types {
                    self.add_schema(variant, false);
                }
            }
            _ => {}
        };

        // When a struct field, the field description overrides the
        // declaration description, so avoid inheriting it!
        if is_field {
            schema.description = None;
        }

        // Store the name so that we can use it as a reference for other types
        if let Some(name) = &schema.name {
            self.references.insert(name.to_owned());

            // Types without a name cannot be rendered at the root
            if !self.schemas.contains_key(name) {
                self.schemas.insert(name.to_owned(), schema);
            }
        }
    }

    /// Generate an output by rendering all collected [`Schema`]s using the provided
    /// [`SchemaRenderer`], and finally write to the provided file path.
    pub fn generate<P: AsRef<Path>, O, R: SchemaRenderer<'gen, O>>(
        &'gen self,
        output_file: P,
        mut renderer: R,
    ) -> miette::Result<()> {
        let output_file = output_file.as_ref();

        let mut output = renderer.render(&self.schemas, &self.references)?;
        output.push('\n');

        if let Some(parent) = output_file.parent() {
            fs::create_dir_all(parent).into_diagnostic()?;
        }

        fs::write(output_file, output).into_diagnostic()?;

        Ok(())
    }
}
