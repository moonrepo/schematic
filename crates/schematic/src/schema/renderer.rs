use indexmap::IndexMap;
use schematic_types::*;
use std::collections::HashSet;

pub type RenderResult<T = String> = miette::Result<T>;

/// Renders [`SchemaType`]s to a distinct format (derived from generic `O`)
/// for use within a [`SchemaGenerator`].
pub trait SchemaRenderer<'gen, O = String> {
    /// Return true of the provided name is a referenced type.
    fn is_reference(&self, name: &str) -> bool;

    /// Render an [`ArrayType`] to a string.
    fn render_array(&mut self, array: &ArrayType, schema: &Schema) -> RenderResult<O>;

    /// Render a boolean type to a string.
    fn render_boolean(&mut self, boolean: &BooleanType, schema: &Schema) -> RenderResult<O>;

    /// Render an [`EnumType`] to a string.
    fn render_enum(&mut self, enu: &EnumType, schema: &Schema) -> RenderResult<O>;

    /// Render a [`FloatType`] to a string.
    fn render_float(&mut self, float: &FloatType, schema: &Schema) -> RenderResult<O>;

    /// Render an [`IntegerType`] to a string.
    fn render_integer(&mut self, integer: &IntegerType, schema: &Schema) -> RenderResult<O>;

    /// Render a [`LiteralType`] to a string.
    fn render_literal(&mut self, literal: &LiteralType, schema: &Schema) -> RenderResult<O>;

    /// Render a null type to a string.
    fn render_null(&mut self, schema: &Schema) -> RenderResult<O>;

    /// Render an [`ObjectType`] to a string.
    fn render_object(&mut self, object: &ObjectType, schema: &Schema) -> RenderResult<O>;

    /// Render a referenced type to a string.
    fn render_reference(&mut self, reference: &str, schema: &Schema) -> RenderResult<O>;

    /// Render a [`StringType`] to a string.
    fn render_string(&mut self, string: &StringType, schema: &Schema) -> RenderResult<O>;

    /// Render a [`StructType`] to a string.
    fn render_struct(&mut self, structure: &StructType, schema: &Schema) -> RenderResult<O>;

    /// Render a [`TupleType`] to a string.
    fn render_tuple(&mut self, tuple: &TupleType, schema: &Schema) -> RenderResult<O>;

    /// Render a [`UnionType`] to a string.
    fn render_union(&mut self, uni: &UnionType, schema: &Schema) -> RenderResult<O>;

    /// Render an unknown type to a string.
    fn render_unknown(&mut self, schema: &Schema) -> RenderResult<O>;

    /// Render all possible variants of the provided [`Schema`] to a string.
    /// If a variant has an explicit name, and that name is a reference, return
    /// the name instead of rendering the type.
    fn render_schema(&mut self, schema: &Schema) -> RenderResult<O> {
        if let Some(name) = &schema.name {
            if self.is_reference(name) {
                return self.render_reference(name, schema);
            }
        }

        self.render_schema_without_reference(schema)
    }

    /// Like [`render_schema`] but does not check for references.
    fn render_schema_without_reference(&mut self, schema: &Schema) -> RenderResult<O> {
        match &schema.ty {
            SchemaType::Null => self.render_null(schema),
            SchemaType::Unknown => self.render_unknown(schema),
            SchemaType::Array(array) => self.render_array(array, schema),
            SchemaType::Boolean(boolean) => self.render_boolean(boolean, schema),
            SchemaType::Enum(enu) => self.render_enum(enu, schema),
            SchemaType::Float(float) => self.render_float(float, schema),
            SchemaType::Integer(integer) => self.render_integer(integer, schema),
            SchemaType::Literal(literal) => self.render_literal(literal, schema),
            SchemaType::Object(object) => self.render_object(object, schema),
            SchemaType::Struct(structure) => self.render_struct(structure, schema),
            SchemaType::String(string) => self.render_string(string, schema),
            SchemaType::Tuple(tuple) => self.render_tuple(tuple, schema),
            SchemaType::Union(uni) => self.render_union(uni, schema),
            SchemaType::Reference(name) => self.render_reference(name, schema),
        }
    }

    /// Render the list of [`Schema`]s to a string, in the order they are listed.
    /// References between types can be resolved using the provided `references` set.
    fn render(
        &mut self,
        schemas: &'gen IndexMap<String, Schema>,
        references: &'gen HashSet<String>,
    ) -> RenderResult;
}
