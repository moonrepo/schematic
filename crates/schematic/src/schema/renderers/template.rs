use crate::format::Format;
use crate::schema::{RenderResult, SchemaRenderer};
use indexmap::IndexMap;
use miette::miette;
use schematic_types::*;
use std::collections::{HashMap, VecDeque};

/// Options to control the rendered template.
pub struct TemplateOptions {
    /// Include field comments in output.
    pub comments: bool,

    /// List of field names to render but comment out.
    pub comment_fields: Vec<String>,

    /// Default values for each field within the root struct.
    pub default_values: HashMap<String, SchemaType>,

    /// List of array and object field names to expand and render a fake item.
    pub expand_fields: Vec<String>,

    /// Content to append to the bottom of the output.
    pub footer: String,

    /// Content to prepend to the top of the output.
    pub header: String,

    /// List of field names to not render.
    pub hide_fields: Vec<String>,

    /// Character(s) to use for indentation.
    pub indent_char: String,

    /// Insert an extra newline between fields.
    pub newline_between_fields: bool,
}

impl Default for TemplateOptions {
    fn default() -> Self {
        Self {
            comments: true,
            comment_fields: vec![],
            default_values: HashMap::new(),
            expand_fields: vec![],
            footer: String::new(),
            header: String::new(),
            hide_fields: vec![],
            indent_char: "  ".into(),
            newline_between_fields: true,
        }
    }
}

pub fn lit_to_string(lit: &LiteralValue) -> String {
    match lit {
        LiteralValue::Bool(inner) => inner.to_string(),
        LiteralValue::F32(inner) => inner.to_string(),
        LiteralValue::F64(inner) => inner.to_string(),
        LiteralValue::Int(inner) => inner.to_string(),
        LiteralValue::UInt(inner) => inner.to_string(),
        LiteralValue::String(inner) => format!("\"{}\"", inner),
    }
}

pub fn is_nested_type(schema: &SchemaType) -> bool {
    match schema {
        SchemaType::Struct(_) => true,
        SchemaType::Union(uni) => {
            if uni.is_nullable() && uni.variants_types.len() == 2 {
                uni.variants_types
                    .iter()
                    .find(|v| !v.is_null())
                    .is_some_and(|v| is_nested_type(v))
            } else {
                false
            }
        }
        _ => false,
    }
}

/// Renders template files from a schema.
#[deprecated = "Use the format specific renderers instead!"]
pub struct TemplateRenderer;

#[allow(deprecated)]
impl TemplateRenderer {
    pub fn new_format(format: Format) -> Box<dyn SchemaRenderer<String>> {
        Self::new(format, TemplateOptions::default())
    }

    #[allow(clippy::new_ret_no_self)]
    pub fn new(format: Format, options: TemplateOptions) -> Box<dyn SchemaRenderer<String>> {
        match format {
            Format::None => unreachable!(),

            #[cfg(feature = "json")]
            Format::Json => Box::new(super::jsonc_template::JsoncTemplateRenderer::new(options)),

            #[cfg(feature = "toml")]
            Format::Toml => Box::new(super::toml_template::TomlTemplateRenderer::new(options)),

            #[cfg(feature = "yaml")]
            Format::Yaml => Box::new(super::yaml_template::YamlTemplateRenderer::new(options)),
        }
    }
}

pub struct TemplateContext {
    pub depth: usize,
    pub options: TemplateOptions,

    format: Format,
    stack: VecDeque<String>,
}

impl TemplateContext {
    pub fn new(format: Format, options: TemplateOptions) -> Self {
        Self {
            depth: 0,
            format,
            options,
            stack: VecDeque::new(),
        }
    }

    pub fn indent(&self) -> String {
        if self.depth == 0 {
            String::new()
        } else {
            self.options.indent_char.repeat(self.depth)
        }
    }

    pub fn gap(&self) -> &str {
        if self.options.newline_between_fields {
            "\n\n"
        } else {
            "\n"
        }
    }

    pub fn create_comment(&self, field: &SchemaField) -> String {
        if !self.options.comments {
            return String::new();
        }

        let mut lines = vec![];
        let indent = self.indent();
        let prefix = self.get_comment_prefix();

        let mut push = |line: String| {
            lines.push(format!("{indent}{prefix}{}", line));
        };

        if let Some(comment) = &field.description {
            comment
                .trim()
                .split('\n')
                .for_each(|c| push(c.trim().to_owned()));
        }

        if let Some(deprecated) = &field.deprecated {
            push(if deprecated.is_empty() {
                "@deprecated".into()
            } else {
                format!("@deprecated {}", deprecated)
            });
        }

        if let Some(env_var) = &field.env_var {
            push(format!("@envvar {}", env_var));
        }

        if lines.is_empty() {
            return String::new();
        }

        let mut out = lines.join("\n");
        out.push('\n');
        out
    }

    pub fn create_field(&self, field: &SchemaField, property: String) -> String {
        let key = self.get_stack_key();

        format!(
            "{}{}{}{}",
            self.create_comment(field),
            self.indent(),
            if self.options.comment_fields.contains(&key) {
                self.get_comment_prefix()
            } else {
                ""
            },
            property
        )
    }

    pub fn get_comment_prefix(&self) -> &str {
        if self.format.is_json() {
            "// "
        } else {
            "# "
        }
    }

    pub fn get_stack_key(&self) -> String {
        let mut key = String::new();
        let last_index = self.stack.len() - 1;

        for (index, item) in self.stack.iter().enumerate() {
            key.push_str(item);

            if index != last_index {
                key.push('.');
            }
        }

        key
    }

    pub fn is_expanded(&self, key: &String) -> bool {
        self.options.expand_fields.contains(key)
    }

    pub fn is_hidden(&self, field: &SchemaField) -> bool {
        let key = self.get_stack_key();

        field.hidden || self.options.hide_fields.contains(&key)
    }

    pub fn push_stack(&mut self, name: &str) {
        self.stack.push_back(name.to_owned());
    }

    pub fn pop_stack(&mut self) {
        self.stack.pop_back();
    }
}

pub fn render_array(_array: &ArrayType) -> RenderResult {
    Ok("[]".into())
}

pub fn render_boolean(boolean: &BooleanType) -> RenderResult {
    if let Some(default) = &boolean.default {
        return Ok(lit_to_string(default));
    }

    Ok("false".into())
}

pub fn render_enum(enu: &EnumType) -> RenderResult {
    if let Some(index) = &enu.default_index {
        if let Some(value) = enu.values.get(*index) {
            return Ok(lit_to_string(value));
        }
    }

    render_null()
}

pub fn render_float(float: &FloatType) -> RenderResult {
    if let Some(default) = &float.default {
        return Ok(lit_to_string(default));
    }

    Ok("0.0".into())
}

pub fn render_integer(integer: &IntegerType) -> RenderResult {
    if let Some(default) = &integer.default {
        return Ok(lit_to_string(default));
    }

    Ok("0".into())
}

pub fn render_literal(literal: &LiteralType) -> RenderResult {
    if let Some(value) = &literal.value {
        return Ok(lit_to_string(value));
    }

    render_null()
}

pub fn render_null() -> RenderResult {
    Ok("null".into())
}

pub fn render_object(_object: &ObjectType) -> RenderResult {
    Ok("{}".into())
}

pub fn render_reference(reference: &str) -> RenderResult {
    Ok(reference.into())
}

pub const EMPTY_STRING: &str = "\"\"";

pub fn render_string(string: &StringType) -> RenderResult {
    if let Some(default) = &string.default {
        return Ok(lit_to_string(default));
    }

    Ok(EMPTY_STRING.into())
}

pub fn render_struct(_structure: &StructType) -> RenderResult {
    Ok("{}".into())
}

pub fn render_tuple(
    tuple: &TupleType,
    mut render: impl FnMut(&SchemaType) -> RenderResult,
) -> RenderResult {
    let mut items = vec![];

    for item in &tuple.items_types {
        items.push(render(item)?);
    }

    Ok(format!("[{}]", items.join(", ")))
}

pub fn render_union(
    uni: &UnionType,
    mut render: impl FnMut(&SchemaType) -> RenderResult,
) -> RenderResult {
    if let Some(index) = &uni.default_index {
        if let Some(variant) = uni.variants_types.get(*index) {
            return render(variant);
        }
    }

    // We have a nullable type, so render the non-null value
    if uni.is_nullable() {
        if let Some(variant) = uni.variants_types.iter().find(|v| !v.is_null()) {
            return render(variant);
        }
    }

    render_null()
}

pub fn render_unknown() -> RenderResult {
    render_null()
}

pub fn validate_root(schemas: &IndexMap<String, SchemaType>) -> miette::Result<StructType> {
    let Some(schema) = schemas.values().last() else {
        return Err(miette!(
            "At least 1 schema is required to generate a template."
        ));
    };

    let SchemaType::Struct(root) = schema else {
        return Err(miette!("The last registered schema must be a struct type."));
    };

    Ok(root.to_owned())
}
