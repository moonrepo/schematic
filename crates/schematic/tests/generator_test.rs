#![allow(dead_code, deprecated)]

use schematic::schema::template::TemplateOptions;
use schematic::schema::SchemaGenerator;
use schematic::*;
use starbase_sandbox::{assert_snapshot, create_empty_sandbox};
use std::collections::HashMap;
use std::fs;
use std::path::PathBuf;

derive_enum!(
    /** Docblock comment. */
    #[derive(ConfigEnum, Default)]
    pub enum BasicEnum {
        #[default]
        Foo,
        Bar,
        Baz,
    }
);

/// Some comment.
#[derive(Clone, Config)]
pub struct AnotherConfig {
    /// An optional string.
    opt: Option<String>,
    /// An optional enum.
    enums: Option<BasicEnum>,
}

#[derive(Clone, Config)]
struct GenConfig {
    boolean: bool,
    string: String,
    number: usize,
    float32: f32,
    float64: f64,
    vector: Vec<String>,
    map: HashMap<String, u64>,
    enums: BasicEnum,
    #[setting(nested)]
    nested: AnotherConfig,

    // Types
    date: chrono::NaiveDate,
    datetime: chrono::NaiveDateTime,
    decimal: rust_decimal::Decimal,
    time: chrono::NaiveTime,
    path: PathBuf,
    rel_path: relative_path::RelativePathBuf,
    url: Option<url::Url>,
    version: Option<semver::Version>,
    version_req: semver::VersionReq,
    spec: version_spec::VersionSpec,
    spec_unresolved: version_spec::UnresolvedVersionSpec,
    id: warpgate::Id,
    locator: Option<warpgate::PluginLocator>,
    json_value: serde_json::Value,
    toml_value: Option<toml::Value>,
    yaml_value: serde_yaml::Value,
}

/// Some comment.
#[derive(Clone, Config)]
#[config(env_prefix = "ENV_PREFIX_")]
pub struct TwoDepthConfig {
    /// An optional string.
    opt: Option<String>,
    skipped: String,
}

/// Some comment.
#[derive(Clone, Config)]
pub struct OneDepthConfig {
    /// This is another nested field.
    #[setting(nested)]
    two: TwoDepthConfig,
    #[setting(skip)]
    skipped: String,
}

#[derive(Clone, Config)]
struct TemplateConfig {
    /// This is a boolean with a medium length description.
    #[setting(env = "TEMPLATE_BOOLEAN")]
    boolean: bool,
    /// This is a string.
    #[setting(default = "abc")]
    string: String,
    /// This is a number with a long description.
    /// This is a number with a long description.
    number: usize,
    /// This is a float thats deprecated.
    #[deprecated]
    float32: f32,
    /// This is a float.
    #[setting(default = 1.23)]
    float64: f64,
    /// This is a list of strings.
    vector: Vec<String>,
    /// This is a map of numbers.
    map: HashMap<String, u64>,
    /// This is an enum with a medium length description and deprecated.
    #[deprecated = "Dont use enums!"]
    enums: BasicEnum,
    /// This is a nested struct with its own fields.
    #[setting(nested)]
    nested: AnotherConfig,
    /// This is a nested struct with its own fields.
    #[setting(nested)]
    one: OneDepthConfig,
    skipped: String,

    /// This field is testing array expansion.
    #[setting(nested)]
    expand_array: Vec<AnotherConfig>,
    expand_array_primitive: Vec<usize>,

    /// This field is testing object expansion.
    #[setting(nested)]
    expand_object: HashMap<String, AnotherConfig>,
    expand_object_primitive: HashMap<String, usize>,
}

fn create_generator() -> SchemaGenerator {
    let mut generator = SchemaGenerator::default();
    generator.add::<GenConfig>();
    generator
}

fn create_template_generator() -> SchemaGenerator {
    let mut generator = SchemaGenerator::default();
    generator.add::<TemplateConfig>();
    generator
}

fn create_template_options() -> TemplateOptions {
    TemplateOptions {
        comment_fields: vec!["float32".into(), "map".into()],
        expand_fields: vec![
            "expandArray".into(),
            "expandArrayPrimitive".into(),
            "expandObject".into(),
            "expandObjectPrimitive".into(),
        ],
        hide_fields: vec!["skipped".into(), "one.two.skipped".into()],
        ..TemplateOptions::default()
    }
}

#[cfg(feature = "json_schema")]
mod json_schema {
    use super::*;
    use schematic::schema::json_schema::*;

    #[test]
    fn defaults() {
        let sandbox = create_empty_sandbox();
        let file = sandbox.path().join("schema.json");

        create_generator()
            .generate(&file, JsonSchemaRenderer::default())
            .unwrap();

        assert_snapshot!(fs::read_to_string(file).unwrap());
    }
}

#[cfg(all(feature = "template", feature = "json"))]
mod template_json {
    use super::*;
    use schematic::schema::*;

    #[test]
    fn defaults() {
        let sandbox = create_empty_sandbox();
        let file = sandbox.path().join("schema.json");

        create_template_generator()
            .generate(&file, JsoncTemplateRenderer::new(create_template_options()))
            .unwrap();

        assert_snapshot!(fs::read_to_string(file).unwrap());
    }

    #[test]
    fn without_comments() {
        let sandbox = create_empty_sandbox();
        let file = sandbox.path().join("schema.json");

        create_template_generator()
            .generate(&file, JsonTemplateRenderer::new(create_template_options()))
            .unwrap();

        assert_snapshot!(fs::read_to_string(file).unwrap());
    }
}

#[cfg(all(feature = "template", feature = "toml"))]
mod template_toml {
    use super::*;
    use schematic::schema::*;

    #[test]
    fn defaults() {
        let sandbox = create_empty_sandbox();
        let file = sandbox.path().join("schema.toml");

        create_template_generator()
            .generate(&file, TomlTemplateRenderer::new(create_template_options()))
            .unwrap();

        assert_snapshot!(fs::read_to_string(file).unwrap());
    }
}

#[cfg(all(feature = "template", feature = "yaml"))]
mod template_yaml {
    use super::*;
    use schematic::schema::*;

    #[test]
    fn defaults() {
        let sandbox = create_empty_sandbox();
        let file = sandbox.path().join("schema.yaml");

        create_template_generator()
            .generate(&file, YamlTemplateRenderer::new(create_template_options()))
            .unwrap();

        assert_snapshot!(fs::read_to_string(file).unwrap());
    }
}

#[cfg(feature = "typescript")]
mod typescript {
    use super::*;
    use schematic::schema::typescript::*;

    fn generate(options: TypeScriptOptions) -> String {
        let sandbox = create_empty_sandbox();
        let file = sandbox.path().join("types.ts");

        create_generator()
            .generate(&file, TypeScriptRenderer::new(options))
            .unwrap();

        fs::read_to_string(file).unwrap()
    }

    #[test]
    fn defaults() {
        assert_snapshot!(generate(TypeScriptOptions::default()));
    }

    #[test]
    fn enums() {
        assert_snapshot!(generate(TypeScriptOptions {
            enum_format: EnumFormat::Enum,
            ..TypeScriptOptions::default()
        }));
    }

    #[test]
    fn value_enums() {
        assert_snapshot!(generate(TypeScriptOptions {
            enum_format: EnumFormat::ValuedEnum,
            ..TypeScriptOptions::default()
        }));
    }

    #[test]
    fn const_enums() {
        assert_snapshot!(generate(TypeScriptOptions {
            const_enum: true,
            enum_format: EnumFormat::Enum,
            ..TypeScriptOptions::default()
        }));
    }

    #[test]
    fn object_aliases() {
        assert_snapshot!(generate(TypeScriptOptions {
            object_format: ObjectFormat::Type,
            ..TypeScriptOptions::default()
        }));
    }

    #[test]
    fn exclude_refs() {
        assert_snapshot!(generate(TypeScriptOptions {
            exclude_references: vec!["BasicEnum".into(), "AnotherType".into()],
            ..TypeScriptOptions::default()
        }));
    }

    #[test]
    fn external_types() {
        assert_snapshot!(generate(TypeScriptOptions {
            external_types: HashMap::from_iter([(
                "./externals".into(),
                vec!["BasicEnum".into(), "AnotherType".into()]
            )]),
            ..TypeScriptOptions::default()
        }));
    }

    #[test]
    fn no_refs() {
        assert_snapshot!(generate(TypeScriptOptions {
            disable_references: true,
            indent_char: "  ".into(),
            ..TypeScriptOptions::default()
        }));
    }
}
