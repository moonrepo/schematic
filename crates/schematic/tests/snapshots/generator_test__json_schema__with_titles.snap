---
source: crates/schematic/tests/generator_test.rs
expression: "fs::read_to_string(file).unwrap()"
---
{
  "$schema": "http://json-schema.org/draft-07/schema#",
  "type": "object",
  "required": [
    "boolean",
    "date",
    "datetime",
    "decimal",
    "enums",
    "float32",
    "float64",
    "indexmap",
    "indexset",
    "jsonValue",
    "map",
    "nested",
    "number",
    "path",
    "relPath",
    "string",
    "time",
    "tomlValue",
    "url",
    "vector",
    "version",
    "versionReq",
    "yamlValue"
  ],
  "properties": {
    "boolean": {
      "title": "boolean",
      "type": "boolean"
    },
    "date": {
      "title": "date",
      "type": "string",
      "format": "date"
    },
    "datetime": {
      "title": "datetime",
      "type": "string",
      "format": "date-time"
    },
    "decimal": {
      "title": "decimal",
      "type": "string",
      "format": "decimal"
    },
    "enums": {
      "title": "enums",
      "description": "This is a list of `enumerable` values.",
      "default": "foo",
      "allOf": [
        {
          "$ref": "#/definitions/BasicEnum"
        }
      ]
    },
    "float32": {
      "title": "float32",
      "type": "number"
    },
    "float64": {
      "title": "float64",
      "type": "number"
    },
    "indexmap": {
      "title": "indexmap",
      "type": "object",
      "additionalProperties": {
        "type": "string"
      },
      "propertyNames": {
        "type": "string"
      }
    },
    "indexset": {
      "title": "indexset",
      "anyOf": [
        {
          "type": "array",
          "items": {
            "type": "string"
          }
        },
        {
          "type": "null"
        }
      ]
    },
    "jsonValue": {
      "title": "jsonValue",
      "type": [
        "boolean",
        "object",
        "array",
        "number",
        "string",
        "integer"
      ]
    },
    "map": {
      "title": "map",
      "type": "object",
      "additionalProperties": {
        "type": "number"
      },
      "propertyNames": {
        "type": "string"
      }
    },
    "nested": {
      "title": "nested",
      "description": "**Nested** field.",
      "allOf": [
        {
          "$ref": "#/definitions/AnotherConfig"
        }
      ]
    },
    "number": {
      "title": "number",
      "type": "number"
    },
    "path": {
      "title": "path",
      "type": "string",
      "format": "path"
    },
    "relPath": {
      "title": "relPath",
      "type": "string",
      "format": "path"
    },
    "string": {
      "title": "string",
      "type": "string"
    },
    "time": {
      "title": "time",
      "type": "string",
      "format": "time"
    },
    "tomlValue": {
      "title": "tomlValue",
      "anyOf": [
        {
          "type": [
            "boolean",
            "object",
            "array",
            "number",
            "string",
            "integer"
          ]
        },
        {
          "type": "null"
        }
      ]
    },
    "url": {
      "title": "url",
      "anyOf": [
        {
          "type": "string",
          "format": "uri"
        },
        {
          "type": "null"
        }
      ]
    },
    "vector": {
      "title": "vector",
      "description": "This is a list of strings.",
      "type": "array",
      "items": {
        "type": "string"
      }
    },
    "version": {
      "title": "version",
      "anyOf": [
        {
          "type": "string"
        },
        {
          "type": "null"
        }
      ]
    },
    "versionReq": {
      "title": "versionReq",
      "type": "string"
    },
    "yamlValue": {
      "title": "yamlValue",
      "type": [
        "boolean",
        "object",
        "array",
        "number",
        "string",
        "integer"
      ]
    }
  },
  "additionalProperties": false,
  "definitions": {
    "AnotherConfig": {
      "description": "Some comment.",
      "type": "object",
      "required": [
        "enums",
        "opt"
      ],
      "properties": {
        "enums": {
          "title": "enums",
          "description": "An optional enum.",
          "default": "foo",
          "anyOf": [
            {
              "$ref": "#/definitions/BasicEnum"
            },
            {
              "type": "null"
            }
          ]
        },
        "opt": {
          "title": "opt",
          "description": "An optional string.",
          "anyOf": [
            {
              "type": "string"
            },
            {
              "type": "null"
            }
          ]
        }
      },
      "additionalProperties": false
    },
    "BasicEnum": {
      "type": "string",
      "enum": [
        "foo",
        "bar",
        "baz"
      ]
    }
  }
}

